soufi/finders/photon.py

```diff
@@ -1,157 +1,100 @@
-            "or contains(.,'release') "
-            "or contains(.,'extras')]/text()"
-        )
-        for os_dir in os_dirs:
-            bin_url = f"{base_url}{os_dir}"
-            repo = self._get_repo(bin_url)
-            if repo is None:
-                break
-            for package in repo.findall(self.name):
-                if package.evr == self.version:
-                    break
-            else:
-                continue
-            return package.sourcerpm
+        # Walk the tree backwards, so that newer releases get searched first
+        return reversed(tree.xpath('//a/text()'))
+
+    def _get_source_repos(self):
+        dirs = []
+        for release_dir in self._get_dirs():
+            if release_dir[0].isdigit():
+                url = f"{PHOTON_PACKAGES}{release_dir}"
+                try:
+                    content = self._get_url(url)
+                except soufi.exceptions.DownloadError:
+                    continue
+                tree = html.fromstring(content)
+                # Ideally all the SRPM trees would have the exact same
+                # packages in them, but their `aarch64` trees seem to be a
+                # little light.  Prefer x86_64 to be safe
+                dirs += [
+                    url + dir
+                    for dir in tree.xpath(
+                        "//a[text()[contains(.,'srpms')]"
+                        "[contains(.,'x86_64')]]/text()"
+                    )
+                ]
+        return dirs
+
+    def _get_binary_repos(self):
+        """Scrape the Photon repository indexes and find repodata files."""
+        dirs = []
+        # Walk the tree backwards, so that newer releases get searched first
+        for release_dir in self._get_dirs():
+            if release_dir[0].isdigit():
+                url = f"{PHOTON_PACKAGES}{release_dir}"
+                try:
+                    content = self._get_url(url)
+                except soufi.exceptions.DownloadError:
+                    continue
+                tree = html.fromstring(content)
+                dirs += [
+                    url + dir
+                    for dir in tree.xpath(
+                        "//a[text()[not(contains(.,'srpms'))]"
+                        "[contains(.,'x86_64')]]/text()"
+                    )
+                ]
+        return dirs
+
+    def _walk_source_repos(self, name, version=None):
+        # Photon OS does not provide repomd.xml files for their source
+        # repositories, so we need to override the wonderful source lookup
+        # methods with...  this.
+
+        # Short-circuit here to force a binary package lookup in the caller
+        if version is None:
+            return version
+
+        # Re-assemble a source package name, and try to fetch it from all
+        # the source repos.  This is startlingly effective.
+        for repo in self.source_repos:
+            if not repo.endswith('/'):
+                repo += '/'
```

Comment:
I decided instead to do a search-and-destroy of everything doing this, and instead do the "fixup" only in the `YumFinder._get_repo` method (which is cached)

I've been burned a couple of times by trying to use `os.path` functions to manipulate URLs, and have developed a healthy superstition against it, even when I know it will be safe.